rust/src/core/common/converter.rs

```diff
@@ -0,0 +1,1117 @@
+    ) -> Result<Option<Self::Parameters>> {
+        let theta = detect_angle(
+            matrix[(0, 0)].re,
+            matrix[(0, 0)].im,
+            matrix[(1, 1)].re,
+            matrix[(1, 1)].im,
+        );
+        let k = if theta <= 0.0 {
+            0u64
+        } else {
+            (-(theta / PI).log(2.0).round()) as u64
+        };
+        let expected: Matrix = self.construct_matrix(&k)?;
+        if matrix.approx_eq(&expected, epsilon, ignore_global_phase) {
+            Ok(Some(k))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn construct_matrix(&self, k: &Self::Parameters) -> Result<Matrix> {
+        Ok(UnboundGate::PhaseK(*k).into())
+    }
+}
+
+/// Matrix converter object for the R (= IBM U) gate.
+#[derive(Default)]
+pub struct RMatrixConverter {}
+
+impl MatrixConverter for RMatrixConverter {
+    type Parameters = (f64, f64, f64);
+
+    fn detect_matrix(
+        &self,
+        matrix: &Matrix,
+        epsilon: f64,
+        ignore_global_phase: bool,
+    ) -> Result<Option<Self::Parameters>> {
+        let m00 = matrix[(0, 0)];
+        let m01 = matrix[(0, 1)];
+        let m10 = matrix[(1, 0)];
+        let m11 = matrix[(1, 1)];
+
+        let theta = Complex64::new(m00.norm() + m11.norm(), m01.norm() + m10.norm()).arg() * 2.0;
+
+        let phi_phase = try_normalize(m10 * m00.conj());
+        let lambda_phase = if theta < 0.5 * PI {
+            try_normalize(m11 * m00.conj()) * phi_phase.conj()
+        } else {
+            try_normalize(-m01 * m10.conj()) * phi_phase
+        };
+        let lambda = lambda_phase.arg();
+        let phi = phi_phase.arg();
+
+        let theta = if (m10 * m00.conj() * phi_phase.conj()).re < 0.0 {
+            -theta
+        } else {
+            theta
+        };
+
+        let expected: Matrix = self.construct_matrix(&(theta, phi, lambda))?;
+        if matrix.approx_eq(&expected, epsilon, ignore_global_phase) {
+            Ok(Some((theta, phi, lambda)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn construct_matrix(&self, params: &Self::Parameters) -> Result<Matrix> {
+        Ok(UnboundGate::R(params.0, params.1, params.2).into())
+    }
+}
+
+/// Matrix converter object for any matrix of a certain size - simply has the
+/// matrix itself as its parameter type.
+pub struct UMatrixConverter {
+    /// The number of qubits that the matrix should have, or None if don't
+    /// care.
+    num_qubits: Option<usize>,
+}
+
+impl UMatrixConverter {
+    pub fn new(num_qubits: Option<usize>) -> UMatrixConverter {
+        UMatrixConverter { num_qubits }
+    }
+}
+
+impl MatrixConverter for UMatrixConverter {
+    type Parameters = Matrix;
+
+    fn detect_matrix(
+        &self,
+        matrix: &Matrix,
+        _epsilon: f64,
+        _ignore_global_phase: bool,
+    ) -> Result<Option<Self::Parameters>> {
+        if let Some(expected) = self.num_qubits {
+            if matrix.num_qubits().unwrap() != expected {
```

Comment:
This unwrap panics if the `Matrix` returns `None` from `num_qubits` (log2).

Reply:
Maybe better to validate the `num_qubits` field in the contructor of this type.

Reply:
Added check to make test succeed, didn't feel like changing Matrix. After all, it's a matrix, not necessarily a quantum gate matrix; we currently only use it as such but at least the name should be changed then. OTOH maybe we should've called it SquareMatrix.